Fix integer overflow for window ROWS frame

PHILO-HE opened the pull request:
For window ROWS frame, if a large preceding/following value (int32/int64) is used, integer can overflow during the computation for `rawFrameBounds` (int32), which produces unexpected frame and then wrong result.

PHILO-HE commented:
@aditi-pandit, could you please spare some time to review this pr? Thanks!

aditi-pandit commented on velox/exec/Window.cpp:
```diff
@@ -300,8 +305,16 @@ void Window::updateKRowsFrameBounds(
+        (isKPreceding ? -constantOffset : constantOffset) + startRow;
+    for (int i = 0; i < numRows; i++) {
+      // Considers integer overflow case.
+      if (startValue != (int32_t)startValue) {
```
Should we do this check for each row ? Since we know 'i' goes from 0 to numRows, startValue is bounded as well. So we know the value of i where the overflow could trigger. We could avoid the if condition in the loop and write 2 for loops instead.

aditi-pandit commented on velox/exec/Window.cpp:
```diff
@@ -284,8 +284,13 @@ void updateKRowsOffsetsColumn(
-        (startRow + i) + vector_size_t(precedingFactor * offsets[i]);
+    auto startValue = (int64_t)(startRow + i) + precedingFactor * offsets[i];
+    // Considers integer overflow case.
+    if (startValue != (int32_t)startValue) {
```
Encapsulate this in an inline function like isIntegerOverflow or fixBoundForOverflow and re-use the function in both the constant and column overflow cases.

PHILO-HE replied:
The handling for constant preceding/following is different from column now. Maybe, we can just keep this part unchanged. Thanks!

aditi-pandit commented on velox/functions/prestosql/window/tests/AggregateWindowTest.cpp:
```diff
@@ -198,5 +198,80 @@ TEST_F(AggregateWindowTest, testDecimal) {
+      {input}, "count(c1)", overClause, frameClause, expected);
+
+  // Test integer overflow with column-specified preceding (int64).
+  frameClause = "rows between c3 preceding and 0 following";
```
Add a testcase for preceding with int32 overflow. Add a testcase for overflow with a constant value as well.

aditi-pandit commented on velox/functions/prestosql/window/tests/AggregateWindowTest.cpp:
```diff
@@ -198,5 +198,80 @@ TEST_F(AggregateWindowTest, testDecimal) {
+
+  std::string overClause = "partition by c0 order by c1 desc";
+  // Test with literal larger than INT32_MAX (2147483647).
+  std::string frameClause = "rows between 0 preceding and 2147483650 following";
```
Also add a test where the MAX value is hit in the middle of the partition block ? (In previous case all rows have overflow in following)

aditi-pandit reviewed (commented):
Thanks @PHILO-HE. Did a quick round of review.

PHILO-HE commented:
@aditi-pandit, thanks so much for your review!
I just updated this PR.
1. For constant preceding, check whether integer overflow can happen, and if it can, pre-compute the start index begins overflow. If not possible, goes into the original frame bound setting (std::iota). Thus, we can avoid the overflow check for every row.
2. For column preceding/following, I think we have to do the check one row by one row. As it is different from constant, their implementations are still separate.
3. Added more test cases.

PHILO-HE commented:
Hi @aditi-pandit, could you spare some time to review again? Thanks!

PHILO-HE commented:
@mbasmanova, could you review this pr also? Thanks!

aditi-pandit commented on velox/exec/Window.cpp:
```diff
@@ -284,8 +284,25 @@ void updateKRowsOffsetsColumn(
   for (auto i = 0; i < numRows; i++) {
-    rawFrameBounds[i] =
-        (startRow + i) + vector_size_t(precedingFactor * offsets[i]);
+    auto startValue = (int64_t)(startRow + i) + precedingFactor * offsets[i];
```
Nit : 

We can simplify the code a bit

```
rawFrameBounds[i] = (int64_t)(startRow + i) + precedingFactor * offsets[i];
if (rawFrameBounds[i] < INT32_MIN) {
  rawFrameBounds[i] = 0;
} else if (rawFrameBounds[i] > INT32_MAX) {
     // computeValidFrames will replace INT32_MAX set here
     // with partition's final row index.
  rawFrameBounds[i] = INT32_MAX;
}
```

PHILO-HE replied:
Hi @aditi-pandit, thanks for your review!
I just found `rawFrameBounds[i]` is int32. So maybe, we have to use `startValue` (int64) to store intermediate result, then do the checking. Right?

mbasmanova commented:
@aditi-pandit Aditi, thank you for helping to review this PR.

@PHILO-HE Please, ping me when Aditi's comments are addressed.

PHILO-HE commented:
Friendly ping @aditi-pandit, could you please review again? Thanks!

aditi-pandit commented on velox/exec/Window.cpp:
```diff
@@ -284,8 +284,25 @@ void updateKRowsOffsetsColumn(
+}
+
+// Get the start index that integer overflow can happen for kFollowing.
+FOLLY_ALWAYS_INLINE int32_t getOverflowStart(int64_t startValue) {
```
Nit : Maybe call this overflowLimit (instead of overflowStart) and the parameter can be just called k.

This function is only used in one place. We can move this logic within the below function.

PHILO-HE replied:
@aditi-pandit, thanks for your suggestion. As you pointed out, this small piece of code is only used in one place. So I just put it into the caller side, not making it a function or lambda.

aditi-pandit reviewed (commented):
Thanks @PHILO-HE. This code is looking very good. Just had few comments.

aditi-pandit commented:
> Friendly ping @aditi-pandit, could you please review again? Thanks!

Apologize for the delay. I was out for a work conference last week.

PHILO-HE commented on velox/exec/Window.cpp:
```diff
@@ -300,7 +317,29 @@ void Window::updateKRowsFrameBounds(
     auto constantOffset = frameArg.constant.value();
     auto startValue =
-        startRow + (isKPreceding ? -constantOffset : constantOffset);
+        (isKPreceding ? -constantOffset : constantOffset) + startRow;
```
@aditi-pandit, yes, better to explicitly cast it to int64_t. Just updated. Thanks!

PHILO-HE commented on velox/exec/Window.cpp:
```diff
@@ -300,7 +317,29 @@ void Window::updateKRowsFrameBounds(
+        (isKPreceding ? -constantOffset : constantOffset) + startRow;
+
+    if (isKPreceding) {
+      if (startValue < INT32_MIN) {
```
@aditi-pandit, thanks for your comment!
The handling here is based on an assumption, i.e., any window partition's size cannot be larger than INT32_MAX (I also note `rawFrameBounds` is int32_t*). Thus, `startRow - constantOffset` cannot be a positive number when `constantOffset` is so large. So I directly set 0 for all rows here.

aditi-pandit replied:
Ah I see. Maybe rephrase the comment as follows for better clarity
"For overflow in kPreceding frames, k < INT32_MIN. Since the max number of rows in a partition is INT32_MAX, the frameBound will always be bound to the first row of the partition"

PHILO-HE replied:
@aditi-pandit, yes, your suggested comment is better. Just updated. Thanks!

PHILO-HE commented:
> > Friendly ping @aditi-pandit, could you please review again? Thanks!
> 
> Apologize for the delay. I was out for a work conference last week.

@aditi-pandit, no problem. I just updated the code and posted some clarification. Could you help check again? Thanks!

aditi-pandit commented on velox/functions/prestosql/window/tests/AggregateWindowTest.cpp:
```diff
@@ -198,5 +198,124 @@ TEST_F(AggregateWindowTest, testDecimal) {
+  auto c3 = makeFlatVector<int64_t>(
+      {2147483651,
+       2147483650,
+       2147483649,
```
Maybe add one more value < INT32_MAX somewhere in between as well.

PHILO-HE replied:
@aditi-pandit, thanks for your suggestion! Just updated.

aditi-pandit reviewed (commented):
Thanks @PHILO-HE. Have 2 very minor comments. Else this PR is almost finalized.

PHILO-HE commented:
I think I have fixed all comments from Aditi. @mbasmanova, could you review further? Thanks!

mbasmanova reviewed (commented):
@kagamiori Wei, would you help review and merge this fix? Wondering why this issue is not being caught by Fuzzer.

aditi-pandit reviewed (approved):
Thanks @PHILO-HE

PHILO-HE commented:
Hi @kagamiori, could you please review this pr? Thanks!

kagamiori reviewed (approved):
LGTM. Thank you for fixing this bug!

kagamiori commented:
Hi @PHILO-HE, could you rebase this PR onto the latest main? Thanks!

PHILO-HE commented:
> Hi @PHILO-HE, could you rebase this PR onto the latest main? It's needed for committing it. Thanks!

Hi @kagamiori, just rebased. Thanks!
